providers/geode/shedlock-provider-geode/src/main/java/net/javacrumbs/shedlock/provider/geode/GeodeLockProvider.java

```diff
@@ -0,0 +1,128 @@
+
+    private DistributedLockFunction distributedLockFunction = new DistributedLockFunction();
+
+    /**
+     * GeodeLockProvider Constructor
+     * @param clientCache
+     */
+    public GeodeLockProvider(ClientCache clientCache){
+        this.clientCache = clientCache;
+        if(FunctionService.isRegistered(distributedLockFunction.getId())){
+            FunctionService.registerFunction(distributedLockFunction);
+        }
+        log.info(" Function registered {} on Cluster",distributedLockFunction.getId());
+    }
+
+    /**
+     * Attempts to obtained a {@link SimpleLock} by executing a Geode Function on the cluster,
+     * the function only executes once and is fault tolerant to failures.
+     * @param lockConfiguration
+     * @return
+     */
+    @Override
+    public @NotNull Optional<SimpleLock> lock(@NotNull LockConfiguration lockConfiguration) {
+        log.trace("lock - Attempt : {}", lockConfiguration);
+        boolean isLocked = executeFunction(lockConfiguration,Constants.LOCK);
+        if(isLocked){
+            return Optional.ofNullable(new GeodeLock(this,lockConfiguration));
+        }
+        return Optional.empty();
+    }
+
+    private boolean executeFunction(@NotNull LockConfiguration lockConfiguration,String operation) {
+        Execution execution = FunctionService.onServer(this.clientCache.getDefaultPool())
+            .setArguments(new Object[]{ lockConfiguration.getName(),operation,keyLockTime(lockConfiguration) })
+            .withCollector(new DefaultResultCollector());
+        ResultCollector rc = execution.execute(distributedLockFunction);
+        List arrayList = (ArrayList)rc.getResult();
+        Object o = arrayList.get(0);
+        if(o instanceof Boolean){
+            return (boolean) arrayList.get(0);
+        } else if(o instanceof Exception){
+            log.error("lock - Distributed Lock Function Exception {}",o);
+        }
+        return false;
+    }
+
+
+    private long keyLockTime(LockConfiguration lockConfiguration) {
+        Duration between = Duration.between(ClockProvider.now(), lockConfiguration.getLockAtMostUntil());
+        return between.toMillis();
+    }
+
+    /**
+     * Unlocks a {@link GeodeLock} if lease least time is expired ,
+     * else wait for the lease time to expire.
+     * @param lockConfiguration
+     */
+    public void unlock(LockConfiguration lockConfiguration) {
+        String lockName = lockConfiguration.getName();
+        log.trace("unlock - attempt : {}", lockName);
+        final Instant now = ClockProvider.now();
+        final Instant lockAtLeastInstant = lockConfiguration.getLockAtLeastUntil();
+        if (lockAtLeastInstant.isBefore(now)) {
+            boolean isUnlocked = executeFunction(lockConfiguration,Constants.UNLOCK);
+            log.debug("unlock - done : {}", isUnlocked);
+        } else {
+            try {
+                Thread.sleep(Duration.between(now,lockAtLeastInstant).toMillis());
```

Comment:
Would it be possible to do this without blocking current thread?

Reply:
I can , but then unlock would happen in the future and that would have been inconsistent beahviour ( calling app would assume that it is unlocked as it called unlocked ) . Also if i run a async Task then it could throw an error which the calling application won't get . Let me know your thoughts , ideally the leastTime and atMost Behaviour could have been part of the super class which doesnt unlock unless leastTime is passed.

I would ideally want to do away with the leastTime behaviour as systems where the tasks run in millis , the unlock method would unlock only after the task finishes.

Reply:
I admit that the method name is a bit confusing. But other lock providers effectively unlock the task in the future (if `lockConfiguration.getUnlockTime()` is in the future) so this provider should have the same behavior.

Reply:
Ok done